File: OOP/7_Function_Overriding.cpp
#include <bits/stdc++.h>
using namespace std;

class BaseCalculator
{
public:
    // Base class provides the sum method that takes 2 arguments
    virtual int sum(int num1, int num2)
    {
        return num1 + num2;
    }
};

class DerivedCalculator : public BaseCalculator
{
public:
    // Overriding the base class method (2 arguments)
    int sum(int num1, int num2) override
    {
        return num1 + num2 + 100;
    }
};

int main()
{
    DerivedCalculator obj;

    // Calls the overridden sum method in the derived class (2 arguments)
    cout << obj.sum(1, 2) << endl;   // Output: 103 (since it's overridden)

    return 0;
}

File: OOP/7_Function_Overriding.py


# Runtime Polymorphism -> Dynamic Binding | Overriding | Late Binding

# CompileTime Polymorphism-> Static Binding | Overloading | Early Binding

# In Python, method overriding can be achieved when a child class defines a method that is already present in its parent class, but with a different implementation.


# Base class
class Calculator:
    def sum(self, num1, num2):
        return num1 + num2

# Derived class overrides the sum method
class AdvancedCalculator(Calculator):
    def sum(self, num1, num2, num3=None):
        # If a third number is provided, sum all three numbers
        if num3 is not None:
            return num1 + num2 + num3
        # Otherwise, fall back to the parent class's sum method
        return super().sum(num1, num2)

# Create an object of the derived class
cal = AdvancedCalculator()
print(cal.sum(1, 2))      # Output: 3 (from base class)
print(cal.sum(1, 2, 3))   # Output: 6 (overridden method in derived class)
